apollo-federation/src/connectors/json_selection/methods/public/filter.rs

```diff
@@ -0,0 +1,285 @@
+use apollo_compiler::collections::IndexMap;
+use serde_json_bytes::Value as JSON;
+use shape::Shape;
+use shape::ShapeCase;
+use shape::location::SourceId;
+
+use crate::connectors::json_selection::ApplyToError;
+use crate::connectors::json_selection::ApplyToInternal;
+use crate::connectors::json_selection::MethodArgs;
+use crate::connectors::json_selection::VarsWithPathsMap;
+use crate::connectors::json_selection::immutable::InputPath;
+use crate::connectors::json_selection::location::Ranged;
+use crate::connectors::json_selection::location::WithRange;
+use crate::impl_arrow_method;
+
+impl_arrow_method!(FilterMethod, filter_method, filter_shape);
+/// "Filter" is an array method that returns a new array containing only items that match the criteria.
+/// You can use it to filter an array of values based on a boolean condition.
+///
+/// For example, given a selection of [1, 2, 3, 4, 5]:
+///
+/// $->filter(@->eq(3))      result is [3]
+/// $->filter(@->gt(3))      result is [4, 5]
+///
+/// We are taking each value passed into filter via @ and running the condition function against that value.
+/// Only values where the condition returns true will be included in the result array.
+///
+/// Example with objects:
+/// users->filter(@.active->eq(true))    returns only users where active is true
+fn filter_method(
+    method_name: &WithRange<String>,
+    method_args: Option<&MethodArgs>,
+    data: &JSON,
+    vars: &VarsWithPathsMap,
+    input_path: &InputPath<JSON>,
+) -> (Option<JSON>, Vec<ApplyToError>) {
+    let Some(first_arg) = method_args.and_then(|args| args.args.first()) else {
+        return (
+            None,
+            vec![ApplyToError::new(
+                format!("Method ->{} requires one argument", method_name.as_ref()),
+                input_path.to_vec(),
+                method_name.range(),
+            )],
+        );
+    };
+
+    if let JSON::Array(array) = data {
+        let mut output = Vec::new();
+        let mut errors = Vec::new();
+        let mut has_non_boolean_error = false;
+
+        for (i, element) in array.iter().enumerate() {
+            let input_path = input_path.append(JSON::Number(i.into()));
+            let (applied_opt, arg_errors) = first_arg.apply_to_path(element, vars, &input_path);
+            errors.extend(arg_errors);
+
+            match applied_opt {
+                Some(JSON::Bool(true)) => {
+                    output.push(element.clone());
+                }
+                Some(JSON::Bool(false)) | None => {
+                    // Condition is false or errored, exclude the element
+                }
+                Some(_) => {
+                    // Condition returned a non-boolean value, this is an error
```

Comment:
Just want to point out that this works at runtime because these are JSON values and there's no `Unknown` or `Name` type in JSON, only in the shape system.